segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/CacheStorage.java

```diff
@@ -0,0 +1,122 @@
+     *             less than or equal to {@link #getMaxEntryLength()}.
+     * @return An integer representing the address where this data was inserted at. This can be used to invoke {@link #append},
+     * {@link #get}, {@link #delete} or {@link #replace}.
+     * @throws CacheFullException If there is no more capacity in the {@link CacheStorage} to accomodate this entry.
+     */
+    int insert(@NonNull BufferView data);
+
+    /**
+     * Replaces the entry at the given address with a new contents.
+     *
+     * @param address The address to replace. May be obtained via {@link #insert} or {@link #replace}. If there is no
+     *                entry with this address, this method will act like {@link #insert}.
+     * @param data    A {@link BufferView} representing the new contents of the entry. May be empty. Must have
+     *                {@link BufferView#getLength()} less than or equal to {@link #getMaxEntryLength()}.
+     * @return An integer representing the address where this data now resides. Depending on the implementation, this may
+     * or may not be equal to `address`. This can be used to invoke {@link #append}, {@link #get} or {@link #delete}.
+     * @throws CacheFullException If there is no more capacity in the {@link CacheStorage} to accommodate this entry.
+     * @throws CacheException     If another unexpected exception occurred.
+     */
+    int replace(int address, @NonNull BufferView data);
+
+    /**
+     * Gets the number of bytes that can be appended to an entry that has the given length.
+     *
+     * @param currentLength The current length of the entry.
+     * @return An non-negative integer representing the number of bytes that can be appended.
+     */
+    int getAppendableLength(int currentLength);
+
+    /**
+     * Performs an atomic compare-and-appends of data at the end of an entry. This operation will only take effect if
+     * the current length of the entry matches the argument `expectedLength` and both the entry's length and contents
+     * will be updated atomically.
+     *
+     * @param address        An integer representing the address to append to.
+     * @param expectedLength The currently known length of this entry.
+     * @param data           A {@link BufferView} representing the data to append.
+     * @return An integer representing the number of bytes that were appended. If 0, then the entry at the given address
+     * is already at its max capacity (its length equals {@link #getMaxEntryLength()}.
+     * @throws IncorrectCacheEntryLengthException If `expectedLength` does not match the current length of the entry.
+     * @throws IllegalArgumentException If address does not point to a valid entry or the given {@link BufferView} is
+     * too long to fit into the given entry (use {@link #getAppendableLength} to determine maximum length).
+     */
+    int append(int address, int expectedLength, @NonNull BufferView data);
```

Comment:
No unconditional append?

Reply:
see https://github.com/pravega/pravega/pull/4074#discussion_r353981028

The point of `append` is to fill up the last entry in a block. If the upstream code wants more, then it can begin a new entry.

The reason I added this is because blocks are fixed size, so if an entry's length is not a multiple of that size (4KB), then we'll have extra unused memory. Append provides a very fast way of filling up that last block without the need to find new blocks or perform index changes. Appends are 5x faster than inserts as measured by the benchmark I added with this PR.

Reply:
My question is why you are adding only a conditional append operation to the API rather than adding both a conditional and an unconditional one.

Since we are on the topic of appends to fill up blocks, I'm wondering how effective that can be in our case. A block is fairly small (4KB), so I'm wondering how often we are expected to have small enough appends to be able to fill up a block.

Reply:
There is no need for unconditional at this point. It can be extended.

The append situation is not something new. Even in our perf tests we discovered that, for many clients sending many small events (100 bytes), we have a hard(er) time putting all of that into the cache. The same is true with internal Table Segments (for metadata) which store about 100-500 bytes at a time). 

To solve that we either need to create a new entry or retrieve and reinsert an updated entry. Each of these is inefficient for their own reasons.

The other reason we need appends is exactly the 4KB block size. If we only use 200 bytes of it, the remaining 3896 bytes are wasted memory. If we create a new entry for each append, we'd be wasting a lot of it. So append was added also to help solve this (note that Tier 2 reads do not suffer from this problem as they are extended to be multiples of 4KB).

The final reason is that it's much faster to append than to insert. Insert requires locating new Cache Blocks, while Appends do no such thing - they just write to the last block which is conveniently pointed to by the entry's address.

Reply:
The client is expected to batch appends, in which case, even if the individual events are small, the append to the cache is likely to be larger than 4KB. But, the client might not be able to batch and send small append blocks. If that's correct, then the use of appends is optimizing for the latency of low throughout scenarios. For high-throughput scenarios, the appends blocks should be large enough to require new entries.

Is it a fair analysis?

Reply:
Yes, your understanding is correct.

But the memory optimization part plays a heavier weight. Not only does this align the cache entry sizes to the OS's page file size, but it also helps the `ReadIndex` by effectively putting a high bound on the total number of entries (while previously we could stash up to 4096 1 byte entries in 4KB (worst case append scenario), now all of that is represented by a single cache entry).